Increase client robusteness by adding retry and request timeout

steffengr opened the pull request:
Add a request retry logic for connection failures. Use the retry value from the Hadoop configuration if set. This enables the client to not fail in case the NameNode RPC queues are full. The retry logic and its default settings are the same as in the original Java HdfsClient. The calculation of sleep interval was simply ported to Python. Successful requests might be repeated in case of network failures but this is the same in the vanilla Hadoop implementation and the NameNode interfaces are designed to handle this by either being idempotent (see ClientProtocol @Idempotent annotations) or failing upon the second request, e.g. double creations.

Configure the channel socket to use the socket timeout as configured for Hadoop after establishing the connection. This helps to not time out when running slow operations on the NameNode, such as large listings, or when having large and filled up RPC queues.

wouterdebie commented:
Strange that there are conflicts, while I don't see them when I manually merge.

ravwojdyla commented:
@wouterdebie just a thought - I merged #193 after this PR, so it may be that @steffengr might need to `rebase` on current `master` (?)

ravwojdyla commented:
Thanks @steffengr !

Some comments on PR (please let us know what you think):
 * current implementation adds retry logic to `HAClient`, while `Client` could use retry logic as well, thus would be great to implement it for both
 * there are no tests on retry path
 * what are the scenarios when we do not want to retry? For example do we want to retry on host not reachable exception? If standby namenode process/host is down or there is network partition - would it make sense to waste time to retry, thus delay namenode failover (to active one in this case)? 

Possible solutions to issues above:
 * move retry logic to subclass of `RpcService` (underneath `Client`) - for example `RetryRpcService`, which would handle specific exceptions and retry on rpc layer. This way a client can choose to use specific `RpcService` to either allow retries or not - with default set to `RetryRpcService`(?), furthermore this would be consistent across `HAClient` and `Client`. See example of implementing subclass of `RpcService` [here](https://github.com/spotify/snakebite/pull/144/files#diff-c8f057632fb0c363a91cb945e559cab0R67) - in this case `HARpcService` has logic on handling multiple namenodes on rpc layer instead of `Client`. `RetryRpcService would just trap specific errors, retry according to policy and surface exception to Client only if needed. In fact `RetryRpcService` could reuse code from [here](https://github.com/spotify/snakebite/pull/144/files#diff-c8f057632fb0c363a91cb945e559cab0R67) and add retry logic ontop. This would nicely separate client logic from things like failover over namenodes and retry logic (which most probably should not be in client logic). Does this make sense?
 * see example of "reproducing" network issues [here](https://github.com/spotify/snakebite/pull/144/files#diff-9e86501afb4dd0a1986837a2c30ba9efR66), and for example count number of calls (retries) with specific namenode on mock (for example with [assert_has_calls](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_has_calls))

ravwojdyla commented on snakebite/client.py:
```diff
@@ -1425,10 +1476,16 @@ def __handle_socket_error(self, exception):
         if exception.errno in (errno.ECONNREFUSED, errno.EHOSTUNREACH):
             # if NN is down or machine is not available, pass it:
-            pass
+            if self.__should_retry():
```
Does it always make sense to retry here? Might this sometimes delay "failover" to maybe active namenode?

steffengr replied:
You are right, I ignored that point. The Java client behavior is actually to failover directly on the following exceptions: ConnectException, NoRouteToHostException, UnknownHostException, StandbyException and ConnectTimeoutException. It also has a second limit for max failovers which is 15 by default.

Considering that the clients might otherwise freeze when the configured first NN is down. I'll try to implement that behavior.

ravwojdyla commented on snakebite/client.py:
```diff
@@ -1425,10 +1476,16 @@ def __handle_socket_error(self, exception):
         elif isinstance(exception, socket.timeout):
             # if there's communication/socket timeout, pass it:
-            pass
+            if self.__should_retry():
```
Would it be fine to say:

```
if self.__should_retry(): return
```
?

steffengr replied:
I think so.

ravwojdyla commented on snakebite/client.py:
```diff
@@ -1486,7 +1543,7 @@ class AutoConfigClient(HAClient):
         '''
         :param hadoop_version: What hadoop protocol version should be used (default: 9)
-        :type hadoop_version: int
+        :type hadoop_version: int 
```
+space

ravwojdyla commented on test/client_test.py:
```diff
@@ -71,3 +71,4 @@ def test_empty_namenodes_autoclient(self, environ_get):
         HDFSConfig.hdfs_try_paths = ()
         HDFSConfig.core_try_paths = ()
         self.assertRaises(OutOfNNException, AutoConfigClient)
+
```
+white line

ravwojdyla commented on snakebite/client.py:
```diff
@@ -1412,6 +1440,29 @@ def _switch_namenode(self, namenodes):
+            self.retries += 1
+            return True
+
     def __handle_request_error(self, exception):
```
Are there any request exception that we could retry on?

steffengr replied:
There is actually a RetriableException exception that wraps other exceptions and in this case one might retry. However it's barely used. But I'll have a look and try to get it in.

ravwojdyla commented on snakebite/client.py:
```diff
@@ -1425,10 +1476,16 @@ def __handle_socket_error(self, exception):
+                pass
         else:
             raise
         self.namenode.next()
```
Are there any other [socket.error](https://docs.python.org/2/library/socket.html#socket.error) except `timeout`, `errno.ECONNREFUSED` and `errno.EHOSTUNREACH` that we want to retry on?

steffengr replied:
Quite possible. I'll have a look.

steffengr commented:
Thanks for the review Rav. I'll submit an improved version as soon as I find time.

steffengr commented:
I've fixed the failover behavior. It should now be identical to the FailoverOnNetworkExceptionRetry strategy of the Java implementation. It will failover immediately and up to 15 times (configurable) on connection errors and uses an exponential sleep interval in-between. It will use the previously implemented retry strategy in RetriableExceptions from requests. Other IOExceptions will lead to failures for now as some operations have an at most once flag and I want to prevent them from being executed twice.

I've also addressed the smaller comments Rav had.

I did not yet implement a good test case for it but I have an instance running on our cluster that is periodically triggering list operations and I've observed the desired behavior so far.

As this is tightly coupled with the failover, I find it hard to put it in the client. Also as the HAClient can be used as a replacement of the normal Client, I wonder if it would actually make sense?

steffengr commented:
There was one test case failing because of a bug when running out of failover retries. Should be fixed now.

Also I've observed the behavior of the patched snakebite client during a NN failover today that took about 3 minutes and the patched client retried  successfully until the new active was available and continued using it without any issues.

ravwojdyla commented on snakebite/client.py:
```diff
@@ -1444,12 +1514,14 @@ def wrapped(self, *args, **kw):
                     self.__handle_request_error(e)
                 except socket.error as e:
                     self.__handle_socket_error(e)
+
```
Any need for white line?

ravwojdyla commented on snakebite/client.py:
```diff
@@ -1394,49 +1429,84 @@ def __init__(self, namenodes, use_trash=False, effective_user=None, use_sasl=Fal
         log.debug("Request failed with %s" % exception)
         if exception.args[0].startswith("org.apache.hadoop.ipc.StandbyException"):
-            pass
+            self.namenode.next() # Failover and retry until self.max_failovers was reached
```
Smart!

ravwojdyla commented:
Thanks for updates @steffengr - very exciting!

This looks good - to address your question about retry logic in `Client`, it could be moved to subclass of `RpcService`, thus retry logic would be consistent across `Client`s. What do you think @steffengr ?

wouterdebie commented:
Hey guys! What's the status of this?

pragmattica commented:
I'm running into robustness issues along these lines when downloading large numbers of very large files from an HDFS filesystem. I'd really like to see this merged in (if it's ready). Thanks!

wouterdebie commented:
@ravwojdyla @steffengr guys, any update?

steffengr commented:
Sorry for the late answer, I missed out on that. I've been testing this a bit over the last weeks and it seems to work. I've not had the time to deploy it on some nodes or to run more advanced test scenarios. Also, I haven't had the time to write good test cases for it. I would be happy to get some help here.

If it comes to the refactoring, I think I need to pass right now. It would be more important for me to get it tested in the current state.

wouterdebie commented:
So what's the plan? I hear there are quite a few people waiting for this. @ravwojdyla could you and me take some time together to review/refactor?

steffengr commented:
@wouterdebie @ravwojdyla we could sit together some time next week and get it done.

steffengr commented:
I added a bunch of test cases so this PR should be complete.

ravwojdyla commented:
lgtm - will merge when green.
